### PR TITLE
feat(engine): process pending fcu after release of db write hook

### LINKED ISSUE #8203: Buffer incoming engine messages if pruner is active
### Describe the feature

currently, pruning data and engine message handling are mutually exclusive because we can't handle them concurrently due to exclusive db lock.

rn we're dropping messages in edge cases, which is not ideal.
we expect the pruner to be active for at most a few 100ms, so we should be able to delay engine message processing by buffering them in a Vecdeque and popping them in poll loop:

https://github.com/paradigmxyz/reth/blob/a8bbab2470c08140f10d327989a1ac3122ca2211/crates/consensus/beacon/src/engine/mod.rs#L1870-L1870

we can change this to, draining the incoming messages and storing them in a vecque temporarily, and proceed to pop from the vecdque once we're ready to handle them

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

Process pending FCU after release of database write hook. Supersedes #8238. Closes https://github.com/paradigmxyz/reth/issues/8203.

### CODE REVIEW

File: crates/consensus/beacon/src/engine/mod.rs [hunk 1]
```diff
@@ -186,6 +186,16 @@ where
     payload_validator: ExecutionPayloadValidator,
     /// Current blockchain tree action.
     blockchain_tree_action: Option<BlockchainTreeAction<EngineT>>,
+    /// Pending forkchoice update.
+    /// It is recorded if we cannot process the forkchoice update because
+    /// the hook with database read-write access is active.
```
Comment (onbjerg) [inline crates/consensus/beacon/src/engine/mod.rs:191]:
```suggestion
    /// a hook with database read-write access is active.
```

nit (more general)

File: crates/consensus/beacon/src/engine/mod.rs [hunk 2]
```diff
@@ -508,20 +504,37 @@ where
...
-            Ok(None) => {
-                self.set_blockchain_tree_action(
-                    BlockchainTreeAction::MakeForkchoiceHeadCanonical { state, attrs, tx },
-                );
+            Ok(on_updated_result) => {
+                if let Some(on_updated) = on_updated_result {
+                    // Pre-validate forkchoice state update and return if it's invalid
+                    // or cannot be processed at the moment.
+                    self.on_forkchoice_updated_status(state, on_updated, tx);
+                } else if let Some(hook) = self.hooks.active_db_write_hook() {
```
Comment (onbjerg) [inline crates/consensus/beacon/src/engine/mod.rs:512]: why not match arms?

**Review by onbjerg — COMMENTED** [review]
logic makes sense to me, some nits

File: crates/consensus/beacon/src/engine/mod.rs (hunk 2 above)
Comment (rkrasiuk) [reply crates/consensus/beacon/src/engine/mod.rs:512]: no particular reason, there would still be another `if let Some else` in `Ok(None)` arm

**Review by mattsse — APPROVED** [review]
lgtm

File: crates/consensus/beacon/src/engine/mod.rs [hunk 3]
```diff
@@ -508,20 +504,37 @@ where
...
+                    // We can only process new forkchoice updates if no hook with db write is
+                    // running, since it requires exclusive access to the
+                    // database
+                    let replaced_pending =
+                        self.pending_forkchoice_update.replace((state, attrs, tx));
```
Comment (emhane) [inline crates/consensus/beacon/src/engine/mod.rs:517]: you're still dropping the message here @rkrasiuk , reopening issue

**Pull request merged**